### PR TITLE
Implementing basic `--iree-execution-model=async-external` support.

### PR DESCRIPTION
**Pull request by benvanik**
Implementing basic `--iree-execution-model=async-external` support.
Currently only coarse fences are supported: when the flag is specified
exported functions will take a wait and signal fence pair. Upon return to
a caller execution is not assumed to have completed and the caller can
either wait on the signal fence or chain further invocations with it.
Future invocation models will support specifying an arbitrary set of
fences that allow for up to per-I/O granularity (attention layers could
signal sooner than full decoders, etc) and specifying fences for in-place
buffers (wait until buffer is available to write in before filling).

This initial version is conservative and may include additional queue
barriers in order to signal the user-provided fence but future
improvements to timepoint elision and IPO will make that better. Nearly
all models we work with today end up becoming async with the current
heuristics.

iree-run-module/mlir has been updated to support programs compiled with
the async-external mode. iree-benchmark-module now supports pipelined
and concurrent execution via the `--batch_size=` and `--batch_concurrency=`
flags: `batch_size` defines how many invocations there are and
`batch_concurrency` defines how many of those are able to run concurrently.
Using large values isn't yet great as we should be keeping a certain number
of invocations in-flight instead of scheduling all of them at the start of each
benchmark iteration.

Examples:
```
--batch_size=1 --batch_concurrency=1: default single-shot invocation
--batch_size=4 --batch_concurrency=1: 4 sequential invocations
--batch_size=4 --batch_concurrency=2: 2 concurrent sequences of 2 invocations
--batch_size=4 --batch_concurrency=4: 4 concurrent invocations
```

To see how this changes things, resnet50 on 32 cores in the various modes:
```
  1/1: BM_predict/process_time/real_time       34.3 ms          650 ms           20 items_per_second=29.1876/s
  4/1: BM_predict/process_time/real_time       35.8 ms          791 ms           20 items_per_second=27.9287/s
  4/2: BM_predict/process_time/real_time       24.3 ms          625 ms           28 items_per_second=41.1603/s
  4/4: BM_predict/process_time/real_time       20.2 ms          615 ms           48 items_per_second=49.4266/s
```
The performance improvement is from utilization improvements with the 1/1 looking like this:
image
While the 4/4 looks like this:
image

It's also possible to see how pipelining can help here with 8 invocations. All 8 invocations complete in ~1.7ms (~900us if the TODO stream ordered allocations were implemented) while letting the device run wait-free - the main thread is returned to the caller before the first conv layer of the first invocation completes:
image
zooming out:
image

### CODE REVIEW

File: compiler/src/iree/compiler/Bindings/Native/Transforms/WrapEntryPoints.cpp [hunk 1]
```diff
@@ -154,19 +141,66 @@ static LogicalResult wrapImportFunc(mlir::ModuleOp moduleOp,
   return success();
 }
 
+// Populates attributes on |wrapperOp| to support runtime reflection.
+static void populateReflectionAttrs(IREE::ABI::InvocationModel invocationModel,
+                                    func::FuncOp exportOp,
+                                    func::FuncOp wrapperOp) {
+  auto *context = exportOp.getContext();
+  SmallVector<NamedAttribute, 4> attrs;
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/Bindings/Native/Transforms/WrapEntryPoints.cpp:149]: Can you add a comment here that we are assembling exported reflection attrs? I had to read this a couple of times and misread it as assembling new function attrs first.

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/ConvertStreamToHAL.cpp [hunk 2]
```diff
@@ -44,6 +45,62 @@ static Value getOrCreateWaitFence(Location loc, Value timepointFence,
...
+  auto fence = chainOp.getExternalValues().front();
+  if (!fence || !fence.getType().isa<IREE::HAL::FenceType>()) return nullptr;
+
+  // Try really hard to figure out if the fence can be used. A larger analysis
+  // pass running prior to conversion that did some code motion could help
+  // ensure the fence SSA value is usable in the places it is needed - for now
+  // we just do this local check that satisfies most common programs today. IPO
+  // would do something like add the fence as an argument to function calls so
+  // that the functions could consume it but inlining is pretty aggressive now.
+  if (!IREE::Util::isValueUsableForOp(fence, rewriter.getBlock(),
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/ConvertStreamToHAL.cpp:70]: Am I correct in assuming that this precisely covers the usage added in the wrap pass? (Ie. Doing more complicated/ad-hoc use by the user would necessitate precise analysis?

**Review by stellaraccident — APPROVED** [review]
Wow, this is great. Only one brief request. In general, I wish we had a better way to keep track of the cases where we are relying onnon comprehensive analysis that scales to less than general use by users (vs matching the kind of programs we generate at this level). But I couldn't figure out how to make that an actionable request.

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/ConvertStreamToHAL.cpp (hunk 2 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/ConvertStreamToHAL.cpp:70]: Mostly - the nice thing about this path is that it's just an optimization: if we can't find a usable fence then we insert a host/device sync before signaling the fence the user provided (enough to meet the API contract). The thing that a more general analysis and sinking of the user-provided fence into call trees would provide is a greater chance of wait-free behavior especially in complex programs with flow control and async global variables. Will be cool, but thankfully not needed to start :)

**Comment by benvanik** [comment]
> I wish we had a better way to keep track of the cases where we are relying onnon comprehensive analysis that scales to less than general use by users

Agreed! The other day I was chatting about patterns assuming static shapes needing some kind of FIXME-like marker to make them easy to spot. Hopefully in these cases it's mostly about optimizations instead of compilation/correctness but finding those is just as important for someone looking to fix a slow model so it'd be good to track them.

File: compiler/src/iree/compiler/Bindings/Native/Transforms/WrapEntryPoints.cpp (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Bindings/Native/Transforms/WrapEntryPoints.cpp:149]: Any specific phrasing you want over what's there?
```c++
// Populates attributes on |wrapperOp| to support runtime reflection.
static void populateReflectionAttrs(IREE::ABI::InvocationModel invocationModel,
                                    func::FuncOp exportOp,
                                    func::FuncOp wrapperOp) {
```
I think this is fairly clear that it's constructing reflection attrs for runtime reflection, but maybe you'd like some examples?

Comment (benvanik) [reply compiler/src/iree/compiler/Bindings/Native/Transforms/WrapEntryPoints.cpp:149]: Changed the wording a bit to mention how to look them up at runtime.

**Pull request merged**